subprojects/build-cache/src/main/java/org/gradle/caching/configuration/internal/DefaultBuildCacheConfiguration.java

```diff
@@ -57,18 +58,24 @@ public BuildCache getLocal() {
 
     @Override
     public <T extends BuildCache> T local(Class<T> type) {
```

Comment:
Shall we deprecate the two `local` methods which take a `Class<T>` and change the remaining method to `local<Action<? super DirectoryBuildCache>`? I think we should at least deprecate `local(Class<T>)`, since you won't be able to register a different local build cache type anymore.

Reply:
You need `local(Class)` or `local(Class, Action)` to get access to a `DirectoryBuildCache` instance, so we can't deprecate it yet.

Since using a custom local cache is only deprecated but not forbidden, we can't upgrade `local(Action<? super BuildCache>)` to `local(Action<? super DirectoryBuildCache>)` either.

We can do both in 6.0, though, when we completely remove the ability to have custom local build cache types.

Reply:
The `local(Class)` method is not required to configure the directory build cache, only the `local(Class, Action)` method, right?

Reply:
Can we already introduce `local(Action<? super DirectoryBuildCache>)`? Then we could deprecate `local(Class, Action)` as well.

Reply:
> The `local(Class)` method is not required to configure the directory build cache, only the `local(Class, Action)` method, right?

There won't be a difference between `local(DirectoryBuildCache.class)` and `getLocal()` in 6.0, so we can deprecate the former in favor of the latter. But in 5.x that's not the case, so if you are using `local(DirectoryBuildCache.class).setDirectory(...)` or something today, we can't tell you to use `getLocal()` instead, not yet.

We also can't introduce `local(Action<? super DirectoryBuildCache>)` because it's still possible to have, say, an `HttpBuildCache` configured as a local cache.

Reply:
I think we can introduce `local(Action<? super DirectoryBuildCache)`. In bytecode, that is `local(Action)` anyway. So the question is what happens at runtime.

If the action works on `BuildCache` as now, then it will pass.

If a user of Gradle 5.4 uses the method to configure `DirectoryBuildCache` and didn't use another local build cache, then it will work, too.

We would currently need to do some ugly casting of the action in the implementation, though that should break users using the current API.

If we have the `local` method, which works on `DirectoryBuildCache`, then we can deprecate the other two methods.

Reply:
FTR, this is the problematic code:

```groovy
local(HttpBuildCache)
local { DirectoryBuildCache cache -> 
    cache.directory = ...
}
```

We also can't deprecated `local(Class)` now without breaking something, because someone might have:

```groovy
local(DirectoryBuildCache).directory = ...
```

We want to redirect them to use this:

```groovy
local.directory = ...
```

But since we can't change `getLocal()`'s return type yet, that doesn't work.

Let's go slow with the upgrade. I think stability is more important for this API than simplifying it. The more important reason why we are doing this is to be able to get rid of a lot of complexity internally in 6.0.